File: ekata/ekata/custom_scripts/stock_entry/stock_entry.py
import frappe
from frappe.model.mapper import get_mapped_doc

def on_submit(doc,method=None):
    for item in doc.items:
        frappe.db.set_value('Stock Ledger Entry',{'voucher_no':doc.name,'item_code':item.item_code},'supplier',item.supplier)

def validate(doc,method=None):
	print('\n\n--------------set basic rate--------------\n\n')
	if doc.stock_entry_type == 'Repack' or doc.stock_entry_type == 'Bulking':
		rm_amount = 0
		fg_qty = 0
		rate = 0
		last_fg_item = []
		for i in doc.items:
			if i.custom_is_process_loss:
				i.db_set('is_finished_item', 0)

			i.db_set('set_basic_rate_manually', 1)
			if not i.is_finished_item and not i.is_scrap_item and not i.custom_is_process_loss:
				rm_amount += i.amount
			if i.is_finished_item:
				print(i.qty)
				fg_qty += i.qty
			if i.custom_is_process_loss:
				i.db_set('basic_rate', 0)
				i.db_set('amount', 0)
		rate = rm_amount/fg_qty
		print('>>> gf qty', fg_qty)
		if rate > 0:
			for i in doc.items:
				if i.is_finished_item == 1:
					last_fg_item.append(i)
					amount = rate * i.qty
					print('>>> rate', rate, amount)
					i.db_set('basic_rate', rate)
					i.db_set('amount', amount)
					i.db_set('basic_amount', amount)

		doc.set_total_incoming_outgoing_value()


@frappe.whitelist()
def create_repack_entry(source_name, target_doc=None):
	print('>>> create repack entry >>>')
	doc = frappe.flags.args.doc
	SEDoc = frappe.get_doc('Stock Entry', doc['name'])
	stock_entry = frappe.new_doc("Stock Entry")
	stock_entry.stock_entry_type = "Repack"
	stock_entry.set_posting_time = 1
	for item in SEDoc.items:
		print('\n\n >>>> ', item)
		if item.is_finished_item == 1:
			stock_entry.append('items', {
				's_warehouse': item.t_warehouse,
				'item_code': item.item_code,
				'qty': item.qty,
				'uom': item.uom,
				'stock_uom': item.stock_uom,
				'basic_rate': item.basic_rate,
				'batch_no': item.batch_no,
				'receipt_no': item.receipt_no,
				'outturn_no': item.outturn_no,
				'season': item.season,
				'coffee_processing_details': item.coffee_processing_details
			})
	return stock_entry

File: ekata/ekata/custom_scripts/stock_entry/stock_entry.js
frappe.ui.form.on('Stock Entry', {
   refresh: function(frm) {
        frm.add_custom_button(__("Custom Stock Ledger"), function() {
                frappe.route_options = {
                    voucher_no: frm.doc.name,
                };
                frappe.set_route("query-report", "Custom Stock Ledger");
        }, __(""));
        frm.add_custom_button(__("Create Repack"), function() {
            frappe.model.open_mapped_doc({
                method: "ekata.ekata.custom_scripts.stock_entry.stock_entry.create_repack_entry",
                frm: frm,
                args: {
                    doc: frm.doc
                },
                run_link_triggers: true
            });
        }, __(""));

        if(frm.doc.docstatus > 0) {
            setTimeout(() => {
                frm.remove_custom_button('Stock Ledger', 'View');
                }, 10);

            cur_frm.add_custom_button(__("New Stock Ledger"), function() {
                frappe.route_options = {
                    voucher_no: frm.doc.name,
                    from_date: frm.doc.posting_date,
                    to_date: moment(frm.doc.modified).format('YYYY-MM-DD'),
                    company: frm.doc.company,
                    show_cancelled_entries: frm.doc.docstatus === 2
                };
                frappe.set_route("query-report", "New Stock Ledger");
            }, __("View"));
        }

   },

   validate: function(frm){
        if(frm.doc.stock_entry_type == 'Repack' || frm.doc.stock_entry_type == 'Bulking'){
            if (frm.doc.items && frm.doc.items.length) {
                var fg_per = 0;
                $.each(frm.doc.items || [], function(i, item) {
                    if (!item.is_finished_item && !item.is_scrap_item && !item.custom_is_process_loss) {
                        console.log(item.qty);
                    } else {
                        console.log(fg_per);
                        fg_per += item.fg_percentage;
                    }
                });
                if (fg_per < 100) {
                    frappe.throw('FG Percentage should be 100%');
                }
            }
        }

        if (frm.doc.outturn_no) {
            frm.events.updateItems(frm, 'outturn_no', frm.doc.outturn_no);
        }
        if (frm.doc.lot_no) {
            frm.events.updateItems(frm, 'receipt_no', frm.doc.lot_no);
        }
   },

   updateItems: function(frm, field, value) {
        if (frm.doc.items && frm.doc.items.length) {
            let doctype = frm.doc.items[0].doctype;
            $.each(frm.doc.items || [], function(i, item) {
                frappe.model.set_value(doctype, item.name, field, value);
            });
        }
   },

   outturn_no: function(frm) {
        if (frm.doc.outturn_no) {
            frm.events.updateItems(frm, 'outturn_no', frm.doc.outturn_no);
        }
   },

   lot_no: function(frm) {
        if (frm.doc.lot_no) {
            frm.events.updateItems(frm, 'receipt_no', frm.doc.lot_no);
        }
   }
});
frappe.ui.form.on('Stock Entry Detail',{
    fg_percentage: function(frm,cdt,cdn){
        var row = locals[cdt][cdn];
        var rm_qty = 0
        var fg_per = 0
        var fg_qty = 0
        if (frm.doc.items && frm.doc.items.length) {
            $.each(frm.doc.items || [], function(i, item) {
                if (!item.is_finished_item && !item.is_scrap_item && !item.custom_is_process_loss)
                {
                    console.log(item.outturn_qty)
                    rm_qty += item.outturn_qty;
                }else{
                    console.log(item.fg_percentage)
                    fg_per += item.fg_percentage;
                }
            });
        }
        console.log('row qty',rm_qty,'fg per', fg_per)
        if (fg_per <= 100 ){
            var fg_qty = (rm_qty * row.fg_percentage) / 100
            fg_qty = Math.round(fg_qty);
            frappe.model.set_value(cdt, cdn, 'qty', fg_qty)
        }else{
            frappe.model.set_value(cdt, cdn, 'fg_percentage', 0)
            frappe.model.set_value(cdt, cdn, 'qty', 0)
            frappe.throw('FG Percentage cannot be greater than 100%');
        }
    }
});